### PR TITLE
update dapp fee split

### PR DESCRIPTION
**Pull request by siovanus**
gas fee 50% to node, 50% to a multisign address
if multisign address is not set, all gas fee goes to node
50% is configurable

### CODE REVIEW

File: smartcontract/service/native/governance/param.go [hunk 1]
```diff
@@ -992,3 +992,23 @@ func (this *ChangeInitPosParam) Deserialize(r io.Reader) error {
 	this.Pos = uint32(pos)
 	return nil
 }
+
+type GasAddress struct {
+	Address common.Address
+}
+
+func (this *GasAddress) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, this.Address[:]); err != nil {
```
Comment (laizy) [inline smartcontract/service/native/governance/param.go]: use `utils.WriteAddress` to consistent with `Deserialize` function

File: smartcontract/service/native/governance/method.go [hunk 2]
```diff
@@ -823,7 +829,25 @@ func executeSplit2(native *native.NativeService, contract common.Address, view u
 	if err != nil {
 		return splitSum, fmt.Errorf("getSplitFee, getSplitFee error: %v", err)
 	}
-	income := balance - splitFee
+
+	//fee split to dapp address
+	balance1 := balance * uint64(100-globalParam2.DappFee) / 100
```
Comment (laizy) [inline smartcontract/service/native/governance/method.go]: seems incompatible with old version. if `DappFee == 0`, old version `income = balance - splitFee`, new version: `income = 0 - splitFee`

File: smartcontract/service/native/governance/method.go [hunk 3]
```diff
@@ -823,7 +829,25 @@ func executeSplit2(native *native.NativeService, contract common.Address, view u
...
+	} else {
+		err := appCallTransferOng(native, utils.GovernanceContractAddress, gasAddress.Address, balance1)
+		if err != nil {
+			return splitSum, fmt.Errorf("appCallTransferOng, appCallTransferOng error: %v", err)
+		}
+	}
+
+	//fee split to node
+	balance2 := balance - balance1
+	income := balance2 - splitFee
```
Comment (laizy) [inline smartcontract/service/native/governance/method.go]: can guarantee not underflow?

File: smartcontract/service/native/governance/governance.go [hunk 4]
```diff
@@ -1627,3 +1634,31 @@ func SetPromisePos(native *native.NativeService) ([]byte, error) {
 
 	return utils.BYTE_TRUE, nil
 }
+
+//set gas address to receive 50% of gas fee
+func SetGasAddress(native *native.NativeService) ([]byte, error) {
+	// get admin from database
+	adminAddress, err := global_params.GetStorageRole(native,
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/governance.go]: operatorAddress?

File: smartcontract/service/native/governance/method.go [hunk 5]
```diff
@@ -828,7 +834,31 @@ func executeSplit2(native *native.NativeService, contract common.Address, view u
...
+	if err != nil {
+		return splitSum, fmt.Errorf("getGasAddress, getGasAddress error: %v", err)
+	}
+	if gasAddress.Address == common.ADDRESS_EMPTY {
+		dappIncome = 0
+	} else {
+		err := appCallTransferOng(native, utils.GovernanceContractAddress, gasAddress.Address, dappIncome)
+		if err != nil {
+			return splitSum, fmt.Errorf("appCallTransferOng, appCallTransferOng error: %v", err)
+		}
```
Comment (Honglei-Cong) [inline smartcontract/service/native/governance/method.go:854]: add dappIncome to splitSum ?

**Review by Honglei-Cong — CHANGES_REQUESTED** [review]

**Review by laizy — APPROVED** [review]

**Pull request merged**